### DISCUSSION TITLE
Type inference in with predefined types

### DISCUSSION THREAD

**Discussion by ColemanDunn** [open-circle/valibot#408] [Q&A] [answered]
Hello! I am wanting to get type inference with my own types I have already defined. Using the example on the readme as a reference:

```ts
import { email, minLength, object, type Output, parse, string } from 'valibot'; // 1.2 kB

// Create login schema with email and password
const LoginSchema = object({
  email: string([email()]),
  password: string([minLength(8)]),
});

// Infer output TypeScript type of login schema
type LoginData = Output<typeof LoginSchema>; // { email: string; password: string }

// Returns data as { email: string; password: string }
parse(LoginSchema, { email: 'jane@example.com', password: '12345678' });
```

I clearly now have the type `LoginData` to use in functions and whatnot, but what if I have defined my own types ahead of time and want those to be the inferred type that could also help me define my schema? For example, if I have the type:
```ts
type MyLoginType = {email: string, password: string}
```

I want to know if I could do something along the lines of (this is just pseudo-code idea):
```ts
const LoginSchema = object< MyLoginType >({
  email: string([email()]),
  password: string([minLength(8)]),
});
```

Then I could get the following:
```ts
const loginData = parse(LoginSchema, SomeObject);
```

And this would result in `loginData` being of type `MyLoginType`. 

Taking it a step further, it could help with the definition of my schema like I mentioned before like so:
```ts
const LoginSchema = object< MyLoginType >({
  email: string([email()]),
  password: number(), // <------------Some TypeScript error here
});
```

Is anything like this currently possible? 
Thanks!

**Comment by fabian-hiller** [comment, marked as answer]
Thank you for creating this discussion. I would write it like this:

```ts
import * as v from 'valibot';

type LoginData = {
  email: string;
  password: string;
};

const LoginSchema = v.object({
  email: v.string([v.email()]),
  password: v.string([v.minLength(8)]),
}) satisfies v.BaseSchema<LoginData>;
```

**Comment by ColemanDunn** [comment]
Thank you!
